internal-api/src/main/java/datadog/trace/api/TagMap.java

```diff
@@ -0,0 +1,2979 @@
+    } else if (bucket instanceof Entry) {
+      Entry existingEntry = (Entry) bucket;
+      if (existingEntry.matches(newEntry.tag)) {
+        thisBuckets[bucketIndex] = newEntry;
+
+        // replaced existing entry - no size change
+        return existingEntry;
+      } else {
+        thisBuckets[bucketIndex] =
+            new BucketGroup(existingEntry.hash(), existingEntry, newHash, newEntry);
+
+        this.size += 1;
+        return null;
+      }
+    } else if (bucket instanceof BucketGroup) {
+      BucketGroup lastGroup = (BucketGroup) bucket;
+
+      BucketGroup containingGroup = lastGroup.findContainingGroupInChain(newHash, newEntry.tag);
+      if (containingGroup != null) {
+        // replaced existing entry - no size change
+        return containingGroup._replace(newHash, newEntry);
+      }
+
+      if (!lastGroup.insertInChain(newHash, newEntry)) {
+        thisBuckets[bucketIndex] = new BucketGroup(newHash, newEntry, lastGroup);
+      }
+      this.size += 1;
+      return null;
+    }
+
+    // unreachable
+    return null;
+  }
+
+  @Override
+  public Entry getAndSet(String tag, Object value) {
+    return this.getAndSet(Entry.newAnyEntry(tag, value));
+  }
+
+  @Override
+  public Entry getAndSet(String tag, CharSequence value) {
+    return this.getAndSet(Entry.newObjectEntry(tag, value));
+  }
+
+  @Override
+  public final TagMap.Entry getAndSet(String tag, boolean value) {
+    return this.getAndSet(Entry.newBooleanEntry(tag, value));
+  }
+
+  @Override
+  public final TagMap.Entry getAndSet(String tag, int value) {
+    return this.getAndSet(Entry.newIntEntry(tag, value));
+  }
+
+  @Override
+  public final TagMap.Entry getAndSet(String tag, long value) {
+    return this.getAndSet(Entry.newLongEntry(tag, value));
+  }
+
+  @Override
+  public final TagMap.Entry getAndSet(String tag, float value) {
+    return this.getAndSet(Entry.newFloatEntry(tag, value));
+  }
+
+  @Override
+  public final TagMap.Entry getAndSet(String tag, double value) {
+    return this.getAndSet(Entry.newDoubleEntry(tag, value));
+  }
+
+  public final void putAll(Map<? extends String, ? extends Object> map) {
+    this.checkWriteAccess();
+
+    if (map instanceof OptimizedTagMap) {
+      this.putAllOptimizedMap((OptimizedTagMap) map);
+    } else {
+      this.putAllUnoptimizedMap(map);
+    }
+  }
+
+  private final void putAllUnoptimizedMap(Map<? extends String, ? extends Object> that) {
+    for (Map.Entry<? extends String, ?> entry : that.entrySet()) {
+      // use set which returns a prior Entry rather put which may box a prior primitive value
+      this.set(entry.getKey(), entry.getValue());
+    }
+  }
+
+  /**
+   * Similar to {@link Map#putAll(Map)} but optimized to quickly copy from one TagMap to another
+   *
+   * <p>For optimized TagMaps, this method takes advantage of the consistent TagMap layout to
+   * quickly handle each bucket. And similar to {@link TagMap#putEntry(Entry)} this method shares
```

Comment:
`putEntry` does not exist anymore, I think you meant
```suggestion
   * quickly handle each bucket. And similar to {@link TagMap#getAndSet(Entry)} this method shares
```